pykokkos/lib/ufuncs.py

```diff
@@ -3009,3 +3009,307 @@ def isfinite(view):
+        # special case defined in API std
+        return view
+    out = pk.View(view.shape, dtype=view.dtype)
+    if len(view.shape) > 3:
+        raise NotImplementedError("only up to 3D views currently supported for ceil() ufunc.")
+        
+    if "double" in str(view.dtype):
+        if view.shape == ():
+            pk.parallel_for(1, ceil_impl_1d_double, view=view, out=out)
+        elif len(view.shape) == 1:
+            pk.parallel_for(view.shape[0], ceil_impl_1d_double, view=view, out=out)
+        elif len(view.shape) == 2:
+            pk.parallel_for(view.shape[0], ceil_impl_2d_double, view=view, out=out)
+        elif len(view.shape) == 3:
+            pk.parallel_for(view.shape[0], ceil_impl_3d_double, view=view, out=out)
+
+    elif "float64" in str(view.dtype):
+        if view.shape == ():
+            pk.parallel_for(1, ceil_impl_1d_double, view=view, out=out)
+        elif len(view.shape) == 1:
+            pk.parallel_for(view.shape[0], ceil_impl_1d_double, view=view, out=out)
+        elif len(view.shape) == 2:
+            pk.parallel_for(view.shape[0], ceil_impl_2d_double, view=view, out=out)
+        elif len(view.shape) == 3:
+            pk.parallel_for(view.shape[0], ceil_impl_3d_double, view=view, out=out)
+
+    elif "float" in str(view.dtype):
+        if view.shape == ():
+            pk.parallel_for(1, ceil_impl_1d_float, view=view, out=out)
+        elif len(view.shape) == 1:
+            pk.parallel_for(view.shape[0], ceil_impl_1d_float, view=view, out=out)
+        elif len(view.shape) == 2:
+            pk.parallel_for(view.shape[0], ceil_impl_2d_float, view=view, out=out)
+        elif len(view.shape) == 3:
+            pk.parallel_for(view.shape[0], ceil_impl_3d_float, view=view, out=out)
+
+    else:
+        raise NotImplementedError
+    return out
+
+@pk.workunit
+def floor_impl_1d_double(tid: int, view: pk.View1D[pk.double], out: pk.View1D[pk.double]):
+    out[tid] = floor(view[tid])
+
+
+@pk.workunit
+def floor_impl_2d_double(tid: int, view: pk.View2D[pk.double], out: pk.View2D[pk.double]):
+    for i in range(view.extent(1)):
+        out[tid][i] = floor(view[tid][i])
+
+
+@pk.workunit
+def floor_impl_3d_double(tid: int, view: pk.View3D[pk.double], out: pk.View3D[pk.double]):
+    for i in range(view.extent(1)):
+        for j in range(view.extent(2)):
+            out[tid][i][j] = floor(view[tid][i][j])
+
+@pk.workunit
+def floor_impl_1d_float(tid: int, view: pk.View1D[pk.float], out: pk.View1D[pk.float]):
+    out[tid] = floor(view[tid])
+
+
+@pk.workunit
+def floor_impl_2d_float(tid: int, view: pk.View2D[pk.float], out: pk.View2D[pk.float]):
+    for i in range(view.extent(1)):
+        out[tid][i] = floor(view[tid][i])
+
+
+@pk.workunit
+def floor_impl_3d_float(tid: int, view: pk.View3D[pk.float], out: pk.View3D[pk.float]):
+    for i in range(view.extent(1)):
+        for j in range(view.extent(2)):
+            out[tid][i][j] = floor(view[tid][i][j])
+
+def floor(view):
```

Comment:
lack of docstring is just the result of me not worrying about the docstrings in the auto-templating code I linked... we could add some into the template

There's also the question of whether it makes sense to commit auto-generated code vs. the stuff that generates it; conversely, committing to a templating engine could be annoying as well. Certainly, NumPy used its own custom templating language for years for generating the ufunc inner loops. I think they're starting to transition to C++ these days, at least in some limited areas.

Reply:
I think we should add the docstring in this PR. Regarding committing the templating engine, I would say not to do that for now. Since we're planning on getting rid of type annotations in the future, writing these kernels should be less tedious in the future.